### ISSUE TITLE
perf: Optimize image layout tracking for the common case where all miplevels/layers are the same layout

### ISSUE THREAD

**Issue by jeffbolznv** [KhronosGroup/Vulkan-ValidationLayers#645]
Performance of array images (and to a lesser extent, mipmap images) in the core validation layer is currently really bad because the layout maps unconditionally track the layout of each (level, layer) pair independently, and each is validated per-draw. We should optimize for the case where the entire image has a single layout.

**Comment by jzulauf-lunarg** [comment]
@jeffbolznv 

**TL;DR**
1) agreed on performance impact, some mitigation present
2) do you have a test case?
3) do you have some sense of the usage mix between "common case" and per subresource layouts?

**Details**
image layout tracking and validation is a performance issue.  It was an area targeted in the last performance revisions about this time last year.   Do you have a test case in particular where you are seeing image layout in the profile, that you could reference/provide (even by private email)?

The "sparse array as map of nested maps" is painfully slow.  I've been initially investigating something like an interval tree to set/get wide swaths, but that's in early prototyping. We hadn't considered a fast path for "all subresources" as it wasn't clear how much gain there would be based on actual usage patterns.

Image layout validation doesn't always occur during each draw call. An initial pass at dirt tracking was done to avoid image layout validation when no layouts have changed.  That could be improved. (there's an image_layout_change_count tracked at command buffer granularity, that could be command_buffer/image granularity)

**Comment by jeffbolznv** [comment]
I did most of my testing with dota2, and a tiny bit on wolfenstein2. I saw quite a few mip mapped images in dota2 but no array images, and there was at least one large array image in wolfenstein2.

I suspect the vast majority of images have the same layout across all subresources. Maybe when doing something like mipmap generation you get differences between levels.

I didn't think through all the details of what data structures would be good. Maybe something like replacing `unordered_map<ImageSubresourcePair, T>` with (kind of hard to read) `unordered_map<VkImage, SmallVector<std::pair<ImageSubresourcePair, T>, 1> >`. "SmallVector<T,N>" is an LLVM type (see also boost::small_vector) that contains static storage for N elements of size T, but can be dynamically resized larger than N. Might also be possible to avoid the ImageResourcePair (i.e. `unordered_map<VkImage, SmallVector<T, 1> >`) and just have a canonical ordering of levels and layers and infer whether it is a subresource from the size of the vector.

**Comment by jzulauf-lunarg** [comment]
Reworked image layout tracking code in command record mode.
